CriteoPublisherSdk/Sources/Logging/CR_LogMessage.m

```diff
@@ -0,0 +1,96 @@
+//
+//  CR_LogMessage.m
+//  CriteoPublisherSdk
+//
+//  Copyright © 2018-2020 Criteo. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "CR_LogMessage.h"
+
+@implementation CR_LogMessage
+
+- (instancetype)initWithTag:(NSString *)tag
+                   severity:(CR_LogSeverity)severity
+                       file:(NSString *)file
+                       line:(NSUInteger)line
+                   function:(NSString *)function
+                  exception:(NSException *_Nullable)exception
+                    message:(NSString *)message {
+  self = [super init];
+  if (self) {
+    _tag = tag;
+    _severity = severity;
+    _file = file;
+    _line = line;
+    _function = function;
+    _timestamp = [NSDate date];
+    _exception = exception;
+    _message = message;
+  }
+
+  return self;
+}
+
++ (instancetype)messageWithTag:(NSString *)tag
```

Comment:
Do you think it's worth covering this method with tests?

Reply:
I don't think so, variadic parameters can only be consumed this way, which is and old are really low level C APIs...

By the way, I could not figure a way to defer that string interpolation at logging time to spare some CPU cycles when the message isn't ultimately logged.
Maybe when working on remote logging I might remove that class altogether to address this, as you cannot store such variadic parameters / interpolate strings natively with arguments as an array.

Maybe you have some ideas on the topic?